grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -51,14 +149,615 @@ class ApiGetOpenApiDescriptionHandler(api_call_handler_base.ApiCallHandler):
+    self.schema_objs: Optional[Dict[str, Schema]] = None
+
+  def _NormalizePathComponent(self, component: str) -> str:
+    """Normalize the given path component to be used in a valid OpenAPI path."""
+    if len(component) > 0 and component[0] == '<' and component[-1] == '>':
+      component = component[1:-1]
+      component = component.split(":")[-1]
+      component = f"{{{component}}}"
+
+    return component
+
+  def _NormalizePath(self, path: str) -> str:
+    """Keep only fixed parts and parameter names from Werkzeug URL patterns.
+
+    The OpenAPI specification requires that parameters are surrounded by { }
+    which are added in _NormalizePathComponent.
+
+    Args:
+      path: The path whose representation will be normalized.
+
+    Returns:
+      The normalized version of the path argument, which is a valid OpenAPI
+      path, with curly brackets around path arguments.
+    """
+    components = path.split("/")
+    normalized_components = [
+      self._NormalizePathComponent(component) for component in components
+    ]
+
+    normalized_path = '/'.join(normalized_components)
+
+    return normalized_path
+
+  def _GetPathArgsFromPath(self, path: str) -> List[str]:
+    """Extract path parameters from a Werkzeug Rule URL."""
+    path_args = []
+
+    components = path.split("/")
+    for component in components:
+      if len(component) > 0 and component[0] == '<' and component[-1] == '>':
+        normalized_component = self._NormalizePathComponent(component)
+        normalized_component = normalized_component[1:-1]
+        path_args.append(normalized_component)
+
+    return path_args
+
+  def _GetTypeName(
+      self,
+      cls: Union[Descriptor, FieldDescriptor, EnumDescriptor, Type, int, str]
+  ) -> str:
+    """Generically extract type name from protobuf Descriptors/Type/int/str."""
+    if isinstance(cls, FieldDescriptor):
+      if cls.message_type:
+        return self._GetTypeName(cls.message_type)
+      if cls.enum_type:
+        return self._GetTypeName(cls.enum_type)
+
+      return self._GetTypeName(cls.type)
+
+    if isinstance(cls, Descriptor):
+      return cls.full_name
+
+    if isinstance(cls, EnumDescriptor):
+      return cls.full_name
+
+    if isinstance(cls, type):
+      return cls.__name__
+
+    if isinstance(cls, int):  # It's a protobuf.Descriptor.type value.
+      return cast(str, primitive_types[cls]["name"])
+
+    return str(cls)  # Cover "BinaryStream" and None.
+
+  def _AddPrimitiveTypesSchemas(self) -> None:
+    """Adds the OpenAPI schemas for Protobuf primitives and BinaryStream."""
+    if self.schema_objs is None:  # Check required by mypy.
+      raise AssertionError(
+        "The container of OpenAPI type schemas is not initialized."
+      )
+
+    primitive_types_schemas = {
+      cast(str, primitive_type["name"]):
+        cast(Dict[str, str], primitive_type["schema"])
+      for primitive_type in primitive_types.values()
+    }
```

Comment:
I would move the `cast`-ed values to a separate variables. Right now this dict comprehension is fairly loaded and hard to read (and it's not even obvious that it is a dict comprehension).

Reply:
Thank you, it was hard to read indeed! I casted the entire `primitive_types_schemas` when using it to update the `self.schema_objs` dictionary. I am not sure if this is what you meant, but please let me know if you know a cleaner way to do the casts.